lib/python/dataset.cpp

```diff
@@ -186,24 +187,22 @@ Returned by :py:func:`DataArray.masks`)");
   Dict of data arrays with aligned dimensions.)");
   options.disable_function_signatures();
   dataset.def(
-      py::init([](const std::map<std::string, std::variant<Variable, DataArray>>
-                      &data,
-                  const std::map<std::string, Variable> &coords) {
+      py::init([](const py::object &data, const py::object &coords) {
         Dataset d;
-        for (auto &&[name, item] : data) {
-          auto visitor = [&d, n = name](auto &object) {
-            d.setData(std::string(n), std::move(object));
-          };
-          std::visit(visitor, item);
+        for (auto &&[name, item] : py::dict(data)) {
+          if (py::isinstance<DataArray>(item)) {
+            d.setData(name.cast<std::string>(), item.cast<DataArray &>());
+          } else {
+            d.setData(name.cast<std::string>(), item.cast<Variable &>());
+          }
         }
-        for (auto &&[dim, coord] : coords)
-          d.setCoord(Dim{dim}, coord);
+        for (auto &&[dim, coord] : py::dict(coords))
+          d.setCoord(Dim{dim.cast<std::string>()}, coord.cast<Variable &>());
```

Comment:
Why is data array above using `to_cpp_map` for the coords but this is not?

Reply:
To avoid extra copies. I don't see a way to pass that map directly to `Dataset`. I would either have to make the map and iterate over it (this would only save us the casts) or construct a `Dict` and move that into the dataste via `setCoords` (this would not reuse the existing function)